server/topic.go

```diff
@@ -1125,32 +1140,53 @@ func (t *Topic) thisUserSub(h *Hub, sess *Session, asUid types.Uid, asLvl auth.L
 			}
 		}
 
-		// Undelete
+		// Undelete.
 		userData.deleted = false
 
 		if isNullValue(private) {
 			private = nil
 		}
 		userData.private = private
 
-		// Add subscription to database
-		sub := &types.Subscription{
-			User:      asUid.String(),
-			Topic:     t.name,
-			ModeWant:  userData.modeWant,
-			ModeGiven: userData.modeGiven,
-			Private:   userData.private,
-		}
+		tname := t.name
+		var sub *types.Subscription
+		if isChanSub {
+			// Check if user is already subscribed.
+			var err error
+			sub, err = store.Subs.Get(pkt.Original, asUid)
+			if err != nil {
+				sess.queueOut(ErrUnknown(pkt.Id, toriginal, now))
+				return changed, err
+			}
+			tname = pkt.Original
+			userData.modeWant = types.ModeCChn
+			oldWant = types.ModeCChn
+			userData.modeGiven = types.ModeCChn
+			oldGiven = types.ModeCChn
+		}
+
+		// Add subscription to database, if missing.
+		if sub == nil {
+			sub = &types.Subscription{
+				User:      asUid.String(),
+				Topic:     tname,
+				ModeWant:  userData.modeWant,
+				ModeGiven: userData.modeGiven,
+				Private:   userData.private,
+			}
 
-		if err := store.Subs.Create(sub); err != nil {
-			sess.queueOut(ErrUnknown(pktID, toriginal, now))
-			return changed, err
-		}
+			if err := store.Subs.Create(sub); err != nil {
+				sess.queueOut(ErrUnknown(pkt.Id, toriginal, now))
+				return changed, err
+			}
 
-		changed = true
+			changed = true
+		}
 
-		// Add user to cache.
-		usersRegisterUser(asUid, true)
+		if !isChanSub {
```

Comment:
Make this an else clause on line 1166?

Reply:
User can be added to cache only when the operation has succeeded. It's not yet known in line 1166 if it has because `store.Subs.Create` may fail.